### PR TITLE
[AST/Module] Provide conveniency functions Module::isClangModule()/isSwiftASTModule()

### PR DESCRIPTION
**Pull request by akyrtzi**

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci smoke test and merge

**Comment by slavapestov** [comment]
@akyrtzi Can you do a quick grep of the codebase to find existing code that could use the new utility functions also?

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -1247,6 +1247,44 @@ bool ModuleDecl::isSystemModule() const {
...
+  Optional<SerializedModuleKind> K;
+
+  for (auto File : M->getFiles()) {
+    switch (File->getKind()) {
+      case FileUnitKind::Source:
+      case FileUnitKind::Builtin:
+      case FileUnitKind::Derived:
+        return SerializedModuleKind::Other;
+      case FileUnitKind::SerializedAST:
+        assert(!K.hasValue() && "cannot handle multi-file modules");
```
Comment (jrose-apple) [inline lib/AST/Module.cpp]: This is really not acceptable. The merge-module build step takes many serialized ASTs and merges them, and the SIL tools accept a SIL source file alongside serialized ASTs.

Comment (jrose-apple) [inline lib/AST/Module.cpp]: I'd really prefer if you didn't ever try to ask "is this a serialized Swift module". It interferes with the idea of someday doing partial rebuilds, or having a non-precompiling script-like import-from-source.

Comment (akyrtzi) [inline lib/AST/Module.cpp]: Are these creating a ModuleDecl object containing mixed FileUnits and/or multiple serialized FileUnits ?
If yes, I could rename the functions to `isSingleClangModule`/`isSingleSwiftASTModule`
A module you will import in source code will be a single serialized AST unit, right ?

Comment (jrose-apple) [inline lib/AST/Module.cpp]: Today, yes. I don't want the system to start assuming this about Swift modules, though.

(The Clang one is fine. At one point I considered having overlays in the same Module as the ClangModuleUnit, but we actually make use of the shadowing behavior we get from today's model.)

Comment (akyrtzi) [inline lib/AST/Module.cpp]: Ok, there are APIs that have this assumption like `ModuleDecl::getModuleFilename()` but I won't add more of them.

**Pull request closed without merging**